### PR TITLE
Iframe api refactor

### PR DESCRIPTION
**Pull request by jonnytest1**

### CODE REVIEW

**Comment by jonnytest1** [comment]
https://github.com/thecodingmachine/workadventure/issues/1059

**Comment by jonnytest1** [comment]
additional information : we should add the test script to the compilation 
because this way it is less obvious if some apis are missing

**Comment by jonnytest1** [comment]
alright seems to work

File: front/src/iframe_api.ts [hunk 1]
```diff
@@ -1,28 +1,61 @@
...
+type WorkadventureFunctions = { [K in ApiKeys]: ObjectWithKeyOfUnion<K> extends Function ? K : never }[ApiKeys]
+
+type WorkadventureFunctionsFilteredByRoot = { [K in WorkadventureFunctions]: ObjectOfKey<K>["addMethodsAtRoot"] extends true ? K : never }[WorkadventureFunctions]
+
+
+type JustMethodKeys<T> = ({ [P in keyof T]: T[P] extends Function ? P : never })[keyof T];
+type JustMethods<T> = Pick<T, JustMethodKeys<T>>;
+
+type SubObjectTypes = {
+    [importCl in WorkadventureCommandClasses as importCl["subObjectIdentifier"]]: JustMethods<importCl>;
```
[the remark this answers is not in the document]
Comment (moufmouf) [reply front/src/iframe_api.ts]: Excellent :D

**Review by moufmouf — CHANGES_REQUESTED** [review]
Hey @jonnytest1 ,

This is really impressive.
I love it.

I have one change to request though, and I don't know if it will be easy (you are mastering Typescript more than I do :) )

You are importing all the files using:

```
const importType = Promise.all([
    import("./Api/iframe/popup"),
    import("./Api/iframe/chatmessage"),
    import("./Api/iframe/Sound"),
    import("./Api/iframe/zone-events"),
    import("./Api/iframe/Navigation"),
    import("./Api/iframe/CoWebsite"),
    import("./Api/iframe/Player"),
    import("./Api/iframe/Bubble")
])
```

These are dynamic imports. As a result, Webpack is generating one JS file per Typescript file.

If you run `yarn run build`, you will see those files in the `dist/` folder:

image

This is a problem because instead of loading one big JS file, we are loading several small ones, and performance-wise, this will slow the startup of the page.

Do you see any efficient way to switch this to a "static" loading?

**Comment by jonnytest1** [comment]
@moufmouf i think this should be one file only (cant really test right now since i get a build error since the new develop version on windows)

File: front/src/Api/iframe/popup.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+
+
+    popupInZone(options: ZonedPopupOptions) {
+        const objectLayerName = options.objectLayerName || options.zone
+
+        let lastOpened = 0;
+
+        let popup: Popup | undefined;
+        zoneCommands.onEnterZone(options.zone, () => {
```
Comment (jonnytest1) [inline front/src/Api/iframe/popup.ts:114]: this one's a bonus ^^

**Comment by moufmouf** [comment]
@jonnytest1 I can confirm this is much better now!

I'll keep the PR open until we find some time to rework the documentation.

Also, I'm not 100% sure I'll keep the `popupInZone` as it is now.
I agree it is a pretty cool utility function, but I wonder if we should not put utility functions in another package that can have a separate life.

Anything we put in the API, we have to maintain for the whole life of WorkAdventure. So I'd rather keep the API as small as possible.

On the other end, there is a huge need to helpers to make the life easier for users. In the future, we could go as far as an helper managing doors (open/close state, etc...)
So maybe I'll remove `popupInZone` from this PR and save it for an external package.

**Comment by jonnytest1** [comment]
> @jonnytest1 I can confirm this is much better now!
> 
> I'll keep the PR open until we find some time to rework the documentation.
> 
> Also, I'm not 100% sure I'll keep the `popupInZone` as it is now.
> I agree it is a pretty cool utility function, but I wonder if we should not put utility functions in another package that can have a separate life.
> 
> Anything we put in the API, we have to maintain for the whole life of WorkAdventure. So I'd rather keep the API as small as possible.
> 
> On the other end, there is a huge need to helpers to make the life easier for users. In the future, we could go as far as an helper managing doors (open/close state, etc...)
> So maybe I'll remove `popupInZone` from this PR and save it for an external package.

maybe we can at least start a community page where scripts and library methods are collected together ?

**Comment by jonnytest1** [comment]
> @jonnytest1 I can confirm this is much better now!
> 
> I'll keep the PR open until we find some time to rework the documentation.
> 
> Also, I'm not 100% sure I'll keep the `popupInZone` as it is now.
> I agree it is a pretty cool utility function, but I wonder if we should not put utility functions in another package that can have a separate life.
> 
> Anything we put in the API, we have to maintain for the whole life of WorkAdventure. So I'd rather keep the API as small as possible.
> 
> On the other end, there is a huge need to helpers to make the life easier for users. In the future, we could go as far as an helper managing doors (open/close state, etc...)
> So maybe I'll remove `popupInZone` from this PR and save it for an external package.

oh we could do an npm package for additional script apis  🤔

i tihnk this one https://github.com/jonnytest1/workadventuremap/blob/master/scripts/conversation.js might also be a good idea 
in one form or another

and maybe an additinoal package that are for apis that require an independent backend server 😁

https://github.com/jonnytest1/workadventuremap/tree/master/scripts loads of stuff in here that might be a good fit

**Review by moufmouf — APPROVED** [review]

**Comment by moufmouf** [comment]
Oops, it seems we are working on the same merge from develop.
Sorry, I'll stop pushing on your PR while you are working on it. Let me know when I can review!

**Comment by jonnytest1** [comment]
> Oops, it seems we are working on the same merge from develop.
> Sorry, I'll stop pushing on your PR while you are working on it. Let me know when I can review!

@moufmouf  ou 😅 sorry 
im finished
i did realize you deleted the play method from sound (or an uncommented version of it  🤔)

File: front/src/Api/iframe/Sound.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        sendToWorkadventure({
+            "type": 'loadSound',
+            "data": {
+                url: this.url,
+            } as LoadSoundEvent
+
+        });
+    }
+
+    public play(config: SoundConfig) {
```
Comment (jonnytest1) [inline front/src/Api/iframe/Sound.ts:18]: should this be removed ?

Comment (moufmouf) [reply front/src/Api/iframe/Sound.ts:18]: Mmmm... no, it is the only way to play a sound right?

```
WA.sound.loadSound('url.mp3).play();
```

Comment (jonnytest1) [reply front/src/Api/iframe/Sound.ts:18]: https://github.com/thecodingmachine/workadventure/commit/fac6574cb5adfffbff8a94f236d1b0cf2cbaad40#diff-b7fd0a9e00605c085d7277dee486fda91141c2a4fa04989151d18691a7cccfc9

well in this commit a playSound method got removed

Comment (jonnytest1) [reply front/src/Api/iframe/Sound.ts:18]: oh but its seems like it was a different method nevermind

**Comment by moufmouf** [comment]
@jonnytest1 Can I work on your branch directly? I'll reorganize the namespaces this way:

```
WA.nav. (goToPage, openTab, goToRoom)
WA.ui. (open/closeBubble, popup functions)
WA.controls. (enable/disableControls)
WA.chat. ...
WA.sound. ...
WA.room.

File: front/src/Api/iframe/Sound.ts (hunk 3 above)
Comment (moufmouf) [reply front/src/Api/iframe/Sound.ts:18]: Yes, it was kind of a shortcut to do `WA.sound.loadSound('url.mp3).play();`
And I'd rather avoid shortcuts and put those in a utility library instead.

**Comment by jonnytest1** [comment]
> @jonnytest1 Can I work on your branch directly? I'll reorganize the namespaces this way:
> 
> ```
> WA.nav. (goToPage, openTab, goToRoom)
> WA.ui. (open/closeBubble, popup functions)
> WA.controls. (enable/disableControls)
> WA.chat. ...
> WA.sound. ...
> WA.room. (zone related functions)
> ```

sure 
PS for new scirpt api contributions its probably best to double check that "addMethodsAtRoot" is not set or false
(or we just remove it from the get go as its only meant for backwards compatibility)

**Pull request merged**